<launch>

  <node pkg="tf2_ros" type="static_transform_publisher" name="bl_imu" args="0 0 0 0 0 0 1 base_link imu_link" />

  <!-- bagfile params -->
  <!-- <arg name="bag_file"        default="$(find ramon_simulation)/bagfiles/rosbot/turtlebot3_house/2020-08-05-22-15-47.bag"/> -->
  <!-- <arg name="bag_file"        default="$(find ramon_simulation)/bagfiles/rosbot/turtlebot3_world/2020-08-07-12-05-25.bag"/> -->
  <!-- <arg name="bag_file"        default="$(find ramon_simulation)/bagfiles/rosbot/turtlebot3_world/2020-08-17-12-06-04.bag"/> -->
  <!-- <arg name="bag_file"        default="$(find ramon_simulation)/bagfiles/rosbot/image_recognition/2020-08-22-15-07-00.bag"/> -->

  <arg name="bag_file"        default="$(find ramon_simulation)/bagfiles/rosbot/turtlebot3_world/2020-08-16-12-15-37.bag"/>
  <!-- <arg name="bag_file"        default="$(find ramon_simulation)/bagfiles/rosbot/ramon_house/2020-11-14-17-51-57.bag"/> -->
  <!-- <arg name="bag_file"        default="$(find ramon_simulation)/bagfiles/rosbot/maze/2020-11-14-17-40-26.bag"/> -->

  <arg name="pcd_save_location" default="$(find ramon_slam3d)/pcd/rosbot/turtlebot3_world/2020-08-16-12-15-37"/>
  <arg name="bag_play"	      default="true"/>
  <arg name="bag_rate"	      default="0.05"/>
      <!-- <arg name="bag_rate"	      default="1.0"/> -->

  <!-- RViz params -->
  <arg name="use_rviz"	      default="true"/>

  <!-- OpenNI params -->
  <arg name="use_openni"	    default="false"/>
  
  <!-- Simulation time (based on bagfile) -->
  <param if="$(arg bag_play)" name="/use_sim_time" value="true"/>

  <arg name="use_ext_odom"    default="false"/>
  <arg name="ext_odom"        default="odom/ekf/enc_imu"/>

  <include if="$(arg use_ext_odom)" file="$(find ramon_slam3d)/launch/robot_localization.launch"/>

  <node if="$(arg bag_play)" 
        pkg="rosbag" 
        type="play" 
        name="rosbag_play_robot"
        args="--clock --rate=$(arg bag_rate) $(arg bag_file)" >
  </node>

  <!-- TFs -->
  <node pkg="tf"
        type="static_transform_publisher"
        name="robot_path_broadcaster"
        args="0 0 0 0 0 0 1 world robot_path3d 100" />        

  <!--  pkg="tf"
        type="static_transform_publisher"
        name="robot_odom_broadcaster"
        args="0 0 0 0 0 0 1 map robot_odom3d 100" />         -->

  <node pkg="tf"
        type="static_transform_publisher"
        name="world_broadcaster"
        args="0 0 0 0 0 0 1 world map 100" />

  <node pkg="tf"
        type="static_transform_publisher"
        name="map2_broadcaster"
        args="0 0 0 0 0 0 1 map odom 100" />

  <node pkg="tf"
        type="static_transform_publisher"
        name="world2_broadcaster"
        args="0 0 0 0 0 0 1 map odom3d 100" />
        
  <node pkg="tf"
        type="static_transform_publisher"
        name="real_odom_broadcaster"
        args="-0.005 0.025 0 0 0 -0.7071068 0.7071068 world real_path 100" />

  <node pkg="tf"
        type="static_transform_publisher"
        name="imu_link_broadcaster"
        args="0 0 0 0 0 0 1 base_link imu_link 10" />

  <node name="ramon_slam3d"	
        pkg="ramon_slam3d"  
        type="slam3d" 
        output="screen"
        clear_params="true">
    <!-- <param name="depth_points_topic" value="/camera/depth/points"/> -->
    <param name="depth_points_topic" value="/camera/depth_registered/points"/>
    <param name="imu_topic" type="string" value="/imu"/>
    <param name="ground_truth_topic" value="/ground_truth/state"/>
    <!-- <param name="pcd_save_location" value="$(arg pcd_save_location)"/> -->
    <param if="$(arg use_ext_odom)" name="extodometry_topic_name" value="$(arg ext_odom)"/>
  </node>

  <!-- Robot -->
  <param name="robot_description" command="$(find xacro)/xacro '$(find ramon_simulation)/urdf/rosbot/rosbot.xacro'"/>

  <!-- RViz -->
    <node if="$(arg use_rviz)"
          name="rviz_OG" 
          pkg="rviz" 
          type="rviz" 
          args="-d $(find ramon_slam3d)/rviz/camera_map_pointcloud_path.rviz" >
    </node>

   <node pkg="robot_localization" type="ekf_localization_node" name="ekf_odom_node" output="screen" >
    <rosparam command="load" file="$(find ramon_slam3d)/config/ekf_localization.yaml"/>
    <!-- <remap from="odometry/filtered" to="odom/ekf/enc_imu"/> -->
   </node>

</launch>